asmcli/components/control-plane/managed.sh

```diff
@@ -24,49 +24,63 @@ call_runIstiod() {
-  fi
-
-  if [[ "${FLEET_ID}" != "${PROJECT_ID}" ]]; then
-    POST_DATA="$(echo "${POST_DATA}" | jq -r --arg MEMBERSHIP "${HUB_IDP_URL/*projects/projects}" '. + {membership: $MEMBERSHIP}')"
-  fi
+    if [[ "${FLEET_ID}" != "${PROJECT_ID}" ]]; then
+      POST_DATA="$(echo "${POST_DATA}" | jq -r --arg MEMBERSHIP "${HUB_IDP_URL/*projects/projects}" '. + {membership: $MEMBERSHIP}')"
+    fi
 
-  info "Provisioning control plane..."
-  retry 2 call_runIstiod "${PROJECT_ID}" "${CLUSTER_LOCATION}" "${CLUSTER_NAME}" "${POST_DATA}"
+    info "Provisioning control plane..."
+    retry 2 call_runIstiod "${PROJECT_ID}" "${CLUSTER_LOCATION}" "${CLUSTER_NAME}" "${POST_DATA}"
 
-  local MUTATING_WEBHOOK_URL
-  MUTATING_WEBHOOK_URL=$(get_managed_mutating_webhook_url)
+    local MUTATING_WEBHOOK_URL
+    MUTATING_WEBHOOK_URL=$(get_managed_mutating_webhook_url)
 
-  local VALIDATION_URL
-  # shellcheck disable=SC2001
-  VALIDATION_URL="$(echo "${MUTATING_WEBHOOK_URL}" | sed 's/inject.*$/validate/g')"
+    local VALIDATION_URL
+    # shellcheck disable=SC2001
+    VALIDATION_URL="$(echo "${MUTATING_WEBHOOK_URL}" | sed 's/inject.*$/validate/g')"
 
-  local CLOUDRUN_ADDR
-  # shellcheck disable=SC2001
-  CLOUDRUN_ADDR=$(echo "${MUTATING_WEBHOOK_URL}" | cut -d'/' -f3)
+    local CLOUDRUN_ADDR
+    # shellcheck disable=SC2001
+    CLOUDRUN_ADDR=$(echo "${MUTATING_WEBHOOK_URL}" | cut -d'/' -f3)
 
-  kpt cfg set asm anthos.servicemesh.controlplane.validation-url "${VALIDATION_URL}"
-  kpt cfg set asm anthos.servicemesh.managed-controlplane.cloudrun-addr "${CLOUDRUN_ADDR}"
+    kpt cfg set asm anthos.servicemesh.controlplane.validation-url "${VALIDATION_URL}"
+    kpt cfg set asm anthos.servicemesh.managed-controlplane.cloudrun-addr "${CLOUDRUN_ADDR}"
 
-  info "Configuring ASM managed control plane revision CRD..."
-  context_append "kubectlFiles" "${CRD_CONTROL_PLANE_REVISION}"
+    info "Configuring ASM managed control plane revision CRD..."
+    context_append "kubectlFiles" "${CRD_CONTROL_PLANE_REVISION}"
 
-  info "Configuring base installation for managed control plane..."
-  context_append "kubectlFiles" "${BASE_REL_PATH}"
+    info "Configuring base installation for managed control plane..."
+    context_append "kubectlFiles" "${BASE_REL_PATH}"
 
-  info "Configuring ASM managed control plane validating webhook config..."
-  context_append "kubectlFiles" "${MANAGED_WEBHOOKS}"
+    info "Configuring ASM managed control plane validating webhook config..."
+    context_append "kubectlFiles" "${MANAGED_WEBHOOKS}"
+  else
+    # `kubectl wait` for non-existent resources will return error directly so we wait in loop
+    info "Wait for the controlplanerevisions CRD to be installed by AFC. This should take a few minutes if cluster is newly registered."
+    for i in {1..10}; do
+      if kubectl wait --for condition=established --timeout=10s crd/controlplanerevisions.mesh.cloud.google.com 2>/dev/null; then
+        break
+      fi
+      sleep 10
+    done
+  fi
```

Comment:
Can you split the if/else logic into separate functions? e.g. `if is_legacy; then install_legacy_mcp; else install_mcp; fi`

Reply:
Done.